x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/Authentication.java

```diff
@@ -117,6 +140,76 @@ public AuthenticationType getAuthenticationType() {
         return metadata;
     }
 
+    /**
+     * Returns a new {@code Authentication}, like this one, but which is compatible with older version nodes.
+     * This is commonly employed when the {@code Authentication} is serialized across cluster nodes with mixed versions.
+     */
+    public Authentication maybeRewriteForOlderVersion(Version olderVersion) {
+        // TODO how can this not be true
+        // assert olderVersion.onOrBefore(getVersion());
+        return new Authentication(
+            getUser(),
+            maybeRewriteRealmRef(olderVersion, getAuthenticatedBy()),
+            maybeRewriteRealmRef(olderVersion, getLookedUpBy()),
+            olderVersion,
+            getAuthenticationType(),
+            maybeRewriteMetadataForApiKeyRoleDescriptors(olderVersion, this)
+        );
+    }
+
+    /**
+     * Returns a new {@code Authentication} that reflects a "run as another user" action under the current {@code Authentication}.
+     * The security {@code RealmRef#Domain} of the resulting {@code Authentication} is that of the run-as user's realm.
+     */
+    public Authentication runAs(User runAs, @Nullable RealmRef lookupRealmRef) {
+        Objects.requireNonNull(runAs);
+        assert false == runAs.isRunAs();
+        assert false == getUser().isRunAs();
+        return new Authentication(
+            new User(runAs, getUser()),
+            getAuthenticatedBy(),
+            lookupRealmRef,
+            getVersion(),
+            getAuthenticationType(),
+            getMetadata()
+        );
+    }
+
+    /** Returns a new {@code Authentication} for tokens created by the current {@code Authentication}, which is used when
+     * authenticating using the token credential.
+     */
+    public Authentication token() {
+        final Authentication newTokenAuthentication = new Authentication(
+            getUser(),
+            getAuthenticatedBy(),
+            getLookedUpBy(),
+            Version.CURRENT,
+            AuthenticationType.TOKEN,
+            getMetadata()
+        );
+        assert Objects.equals(getDomain(), newTokenAuthentication.getDomain());
+        return newTokenAuthentication;
+    }
+
+    /**
+     * Returns {@code true} if the effective user belongs to a realm under a domain.
+     * See also {@link #getDomain()} and {@link #getSourceRealm()}.
+     */
+    public boolean isAssignedToDomain() {
+        return getDomain() != null;
+    }
+
+    /**
+     * Returns the {@link RealmDomain} that the effective user belongs to.
+     * A user belongs to a realm which in turn belongs to a domain.
+     *
+     * The same username can be authenticated by different realms (e.g. with different credential types),
+     * but resources created across realms cannot be accessed unless the realms are also part of the same domain.
+     */
+    public @Nullable RealmDomain getDomain() {
+        return getSourceRealm().getDomain();
```

Comment:
Indeed in some cases (eg when the lookup user does not exist) the user can be a run-as but the lookup `null`.
This is another example of a multitude of inconsistencies in the `Authentication` class.
We need to at least design constructors to prevent such inconsistencies between members of the class.
I will raise an issue about it.